modules/tftp/server.rb

```diff
@@ -95,16 +107,86 @@ def pxeconfig_file(mac)
+        return
+      end
+
+      bootloader_path = "#{bootloader_universe}/pxegrub2/#{os}/#{version}/#{arch}"
+
+      logger.debug "TFTP: Checking bootloader universe for suitable bootloader directory for"
+      logger.debug "TFTP:   * Operating system: #{os}"
+      logger.debug "TFTP:   * Version: #{version}"
+      logger.debug "TFTP:   * Architecture: #{arch}"
+      logger.debug "TFTP: Checking bootloader universe if \"#{bootloader_path}\" exists."
+      unless Dir.exist?(bootloader_path)
+        logger.debug "TFTP: Directory \"#{bootloader_path}\" does not exist."
+
+        bootloader_path = "#{bootloader_universe}/pxegrub2/#{os}/default/#{arch}"
+        logger.debug "TFTP: Checking if fallback directory at \"#{bootloader_path}\" exists."
+        unless Dir.exist?(bootloader_path)
+          logger.debug "TFTP: Directory \"#{bootloader_path}\" does not exist."
+          return
+        end
+      end
+
+      bootloader_path
+    end
+
+    def setup_bootloader(mac, os, major, minor, arch, bootfilename_efi, build)
+      pxeconfig_dir_mac = pxeconfig_dir(mac)
+
+      if build == "true"
+        logger.debug "TFTP: Host is in build mode."
+        logger.debug "TFTP:   => Deploying host specific bootloader files to \"#{pxeconfig_dir_mac}\""
+
+        FileUtils.mkdir_p(pxeconfig_dir_mac)
+
+        version = "#{major}#{".#{minor}" unless minor.empty?}"
+        bootloader_path = bootloader_path(os, version, arch)
+
+        if bootloader_path
+          logger.debug "TFTP: Copying bootloader files from bootloader universe:"
+          logger.debug "TFTP:   - \"#{bootloader_path}/*\" => \"#{pxeconfig_dir_mac}/\""
+          FileUtils.cp_r("#{bootloader_path}/.", "#{pxeconfig_dir_mac}/", remove_destination: true)
+        else
+          logger.debug "TFTP: Copying default bootloader files:"
+          logger.debug "TFTP:   - \"#{pxeconfig_dir}/grub#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/grub#{bootfilename_efi}.efi\""
+          logger.debug "TFTP:   - \"#{pxeconfig_dir}/shim#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/shim#{bootfilename_efi}.efi\""
+          logger.debug "TFTP:   - \"#{pxeconfig_dir}/grub#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/boot.efi\""
+          logger.debug "TFTP:   - \"#{pxeconfig_dir}/shim#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/boot-sb.efi\""
+          FileUtils.cp_r("#{pxeconfig_dir}/grub#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/grub#{bootfilename_efi}.efi", remove_destination: true)
+          FileUtils.cp_r("#{pxeconfig_dir}/shim#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/shim#{bootfilename_efi}.efi", remove_destination: true)
+          FileUtils.cp_r("#{pxeconfig_dir}/grub#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/boot.efi", remove_destination: true)
+          FileUtils.cp_r("#{pxeconfig_dir}/shim#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/boot-sb.efi", remove_destination: true)
+        end
```

Comment:
Here are a couple of points for consideration:
1. It seems unnecessary to call the `pxeconfig_dir` method multiple times. Instead, we can enhance efficiency by storing its value in a variable and reusing it throughout the code.
2. What about scenarios where `grubx64.efi` or `shimx64.efi` are missing from `pxeconfig_dir`, or when there are insufficient permissions to access them?
3. The current implementation may not be suitable for RHEL machines, as it overlooks the necessity of the `shimx64-redhat.efi` file for Secure Boot functionality.

Reply:
Hey Nofar, thanks for your comment.

Regarding topic 1: I agree and I'll change that.

Regarding topic 2:
If the `grubx64.efi` or `shimx64.efi` binary is missing from the `pxeconfig_dir`, provisioning will fail. So if a user wants to use the Bootloader Universe to provide OS specific bootloader files, he has to to stick to the docs and provide all necessary binaries.
Same applies to the permissions: If a user sticks to the docs, the permissions are set correctly.

Regarding topic 3:
This is also a documentation topic: No matter what filename a vendor's Microsoft signed shim binary has, a user has to provide it in the Bootloader Universe and create a symlink `boot-sb.efi` for it. Then everything works as expected.

I had a look at the shim binaries provided by the `shim-x64-***.el8.x86_64.rpm` package of the RHEL 8.8 BaseOS repository with `pesign -S -i`: The `shimx64.efi` binary is signed by Microsoft, the `shimx64-redhat.efi` binary is signed by Red Hat. So my expectation would be, that provisioning a host with Secure Boot turned on should fail with the `shimx64-redhat.efi` binary.

Reply:
AFAIK, for RHEL machines, we require both `shimx64.efi` and `shimx64-redhat.efi` for Secure Boot.
Documentation is necessary to explain that we need to copy them to the `bootloader_universe` directory. However, if we are copying the "default bootloader files" (lines 154-162), `shimx64-redhat.efi` won't be copied.

Reply:
Thanks for the information :) Just be sure I understand the process correctly for RHEL:

1. The boot filename gets set to `shimx64.efi` by Foreman.
2. The `shimx64.efi` calls then the `shimx64-redhat.efi` binary.
3. The `shimx64-redhat.efi`binary finally calls the `grubx64.efi` binary.

Did I get that right?

If so, my current suggestion would be to change the implementation to search for and copy all `shim*.efi` and `grub*.efi` binaries from the `grub2` directory in case we use the default bootloader files. Or should we include the precision (for example `x64`) in the search?
What do you think?

Reply:
I also think that we only need `shimx64.efi` binary which should be signed by MS for default SecureBoot setups.

The `shimx64-<distro>.efi` binaries (other distributions have them too) might be for setups where you provide your own keys in the firmware (PK, KEK, DB). There you can for example explicitly add RH certificate but no MS certificate.

@nofaralfasi can you provide the following output for both RH shims?

```
# pesign -S -i shimx64.efi
# pesign -S -i shimx64-redhat.efi
```

> 2. The shimx64.efi calls then the shimx64-redhat.efi binary.

Cascading of shims do not work AFAIK.

Reply:
After checking, I can confirm that `shimx64-redhat.efi` is not mandatory for SB. I believe it is safe to say that if a user wants to have their own signature (and therefore needs `shimx64-redhat.efi`), it's the user's responsibility to include it in their `bootloader_universe` directory.
So, I think we can drop this issue for now, and if I encounter any problems during my testing, I will bring it up again.